tsl/src/compression/compression.c

```diff
@@ -1106,15 +1106,16 @@ decompress_chunk(Oid in_table, Oid out_table)
 		FreeBulkInsertState(decompressor.bistate);
 	}
 
-	/* Recreate all indexes on out rel, we already have an exclusive lock on it,
-	 * so the strong locks taken by reindex_relation shouldn't matter. */
+	if (reindex_out_table)
```

Comment:
Wasn't the index rebuilt so that it's not inconsistent with the relation? What is the side-effect of not rebuilding the index here? Are we sure the index is not used later in the recompression path?

Reply:
> Wasn't the index rebuilt so that it's not inconsistent with the relation?

Not sure I follow.

Yes, if the index is not rebuild, it might be inconsistent with the relation, but that is not relevant. The point here is that an option is added so that the caller can opt out of rebuilding the index when it is known that it is not necessary to rebuild the index. For example, because it will not be inconsistent (because it is rebuilt before the caller decides to decompress) or because it will later be overwritten (as is the case here). It is entirely up to the caller.

A more relevant question would be if the index needs to be rebuilt or not in each of the cases where this function is called. And in this case we only skip rebuilding the index when doing a decompression as part of a `recompress_chunk`.

> What is the side-effect of not rebuilding the index here?

I'm not sure what side-effects you had in mind. Could you expand a on what you are concerned about here?

> Are we sure the index is not used later in the recompression path?

The index is overwritten as part of the compression (inside `compress_chunk`) by using `reindex_relation` in a similar way to how it is done here. Look in that function if you want to check if the index is used.

Reply:
I think I am confused with the terminology, in_table, out_table etc. By trying to understand the code, it appears to me that `compress_chunk()` reindexes a different relation than `decompress_chunk()`.

Reply:
No, that is right, but there is still no reason to reindex the uncompressed chunk when decompressing since the chunk will be compressed after.

I am still not sure what failure scenario you had in mind.

Reply:
I am looking into whether `compress_chunk()` uses the uncompressed chunk index, which was not re-indexed, in any way. I haven't found anything yet.